twilight-util/src/signature_validation.rs

```diff
@@ -0,0 +1,184 @@
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{self:?}")
+    }
+}
+
+impl std::error::Error for SignatureValidationFailure {}
+
+/// Parsing the public key failed.
+#[derive(Debug)]
+pub enum KeyError {
+    /// The public key was invalid hexadecimal.
+    Hex(FromHexError),
+    /// The public key was malformed.
+    MalformedKey(SignatureError),
+}
+
+impl std::fmt::Display for KeyError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{self:?}")
+    }
+}
+
+impl std::error::Error for KeyError {}
+
+/// The key you are meant to get from the Discord Developer Portal,
+/// on your Application. It is currently listed on the General Information page,
+/// labeled "Public Key", at the time of this writing (January 19th, 2024).
+pub struct Key(VerifyingKey);
+impl Key {
+    /// This function consumes the hexadecimal string which Discord
+    /// provides public keys in. Use `.as_bytes()` on a `&str`, or otherwise
+    /// obtain a byte-string of that text, to use with this function.
+    ///
+    /// # Errors
+    /// This will fail if given invalid hexadecimal, or if the public key fails to
+    /// meet mathematical requirements.
+    pub fn from_hex(pub_key: &[u8]) -> Result<Self, KeyError> {
+        let mut key = [0; 32];
+        hex::decode_to_slice(pub_key, &mut key).map_err(|e| KeyError::Hex(FromHexError(e)))?;
+        VerifyingKey::from_bytes(&key)
+            .map(Self)
+            .map_err(KeyError::MalformedKey)
+    }
+    /// Validate a signature for a given message body, timestamp, and signing key.
+    ///
+    /// (This method is a duplicate of [`check_signature`].)
+    ///
+    /// # Errors
+    /// This will fail if the request being validated has an invalid signature, or if it
+    /// was given the wrong key.
+    pub fn verify(
+        &self,
+        signature: &[u8],
+        timestamp: &[u8],
+        body: &[u8],
+    ) -> Result<(), SignatureValidationFailure> {
+        check_signature(signature, timestamp, body, self)
+    }
+}
+
+/// The name of the HTTP header Discord wants us to read for the signature.
+pub const SIGNATURE_HEADER: &str = "x-signature-ed25519";
+/// The name of the HTTP header Discord wants us to read for the signature timestamp.
+pub const TIMESTAMP_HEADER: &str = "x-signature-timestamp";
+
+/// Validates that a signature is valid for a given message body, timestamp, and signing key.
+///
+/// # Errors
+/// This will fail if the request being validated has an invalid signature, or if it
+/// was given the wrong key.
+pub fn check_signature(
+    signature: &[u8],
+    timestamp: &[u8],
+    body: &[u8],
+    key: &Key,
+) -> Result<(), SignatureValidationFailure> {
+    let mut sig_buf = [0; 64];
+    hex::decode_to_slice(signature, &mut sig_buf)
+        .map_err(|e| SignatureValidationFailure::Hex(FromHexError(e)))?;
+    let sig = Signature::from_bytes(&sig_buf);
```

Comment:
Let's create a `Signature` newtype with a fallible constructor (like `Key`, with an associated `SignatureParseError`) and hide the `ed25519_dalek` dependency from our API.